programs/marginfi/src/instructions/marginfi_account/withdraw.rs

```diff
@@ -55,37 +55,41 @@ pub fn lending_account_withdraw<'info>(
 
     let withdraw_all = withdraw_all.unwrap_or(false);
     let mut marginfi_account = marginfi_account_loader.load_mut()?;
-
-    let in_liquidation = marginfi_account.get_flag(ACCOUNT_IN_RECEIVERSHIP);
     check!(
         !marginfi_account.get_flag(ACCOUNT_DISABLED),
         MarginfiError::AccountDisabled
     );
 
-    let bank_key = bank_loader.key();
-    let maybe_bank_mint;
-
     {
-        let bank = bank_loader.load()?;
+        let mut group = marginfi_group_loader.load_mut()?;
+        let mut bank = bank_loader.load_mut()?;
 
-        maybe_bank_mint =
+        validate_bank_state(&bank, InstructionKind::FailsInPausedState)?;
+
+        let maybe_bank_mint =
             utils::maybe_take_bank_mint(&mut ctx.remaining_accounts, &bank, token_program.key)?;
 
-        if in_liquidation {
-            // Note: we don't care about the price we are just validating non-zero...
-            fetch_asset_price_for_bank(&bank_key, &bank, &clock, ctx.remaining_accounts)?;
-        }
-        validate_bank_state(&bank, InstructionKind::FailsInPausedState)?;
-    } // release immutable borrow of bank
+        let in_receivership = marginfi_account.get_flag(ACCOUNT_IN_RECEIVERSHIP);
+        let price = if in_receivership {
+            let price = fetch_asset_price_for_bank(
+                &bank_loader.key(),
+                &bank,
+                &clock,
+                ctx.remaining_accounts,
+            )?;
 
-    {
-        let group = &marginfi_group_loader.load()?;
+            // Validate price is non-zero during liquidation/deleverage to prevent exploits with stale oracles
+            check!(price > I80F48::ZERO, MarginfiError::ZeroAssetPrice);
 
-        let mut bank = bank_loader.load_mut()?;
+            price
+        } else {
+            // TODO: force users to always pass the oracle, even in case of withdraw_all, to correctly update withdrawn equity.
```

Comment:
I suppose if we implement this on a per bank basis we can actually let this window be denominated in tokens instead so we dont need price, that's one alternative anyways.